File: Core/input_guard.py
"""
Core/input_guard.py — Input validation for RAG queries.

Ported from main-app (George branch) E-12 input length guard.
Validates query length, detects basic prompt-injection patterns,
and ensures Unicode safety for Romanian/multilingual text.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional


# ─────────────────────────────────────────────────────────────────────────────
# Configuration
# ─────────────────────────────────────────────────────────────────────────────

MAX_QUERY_LENGTH = 8000  # characters (matches main-app E-12)

# Patterns commonly used in prompt-injection attempts
_INJECTION_PATTERNS = [
    re.compile(r"ignore\s+(all\s+)?previous\s+instructions", re.IGNORECASE),
    re.compile(r"disregard\s+(all\s+)?above", re.IGNORECASE),
    re.compile(r"you\s+are\s+now\s+(a|an)\s+", re.IGNORECASE),
    re.compile(r"system\s*:\s*", re.IGNORECASE),
    re.compile(r"\[INST\]|\[/INST\]|<\|im_start\|>|<\|im_end\|>", re.IGNORECASE),
]


# ─────────────────────────────────────────────────────────────────────────────
# Result type
# ─────────────────────────────────────────────────────────────────────────────

@dataclass
class ValidationResult:
    valid: bool
    reason: Optional[str] = None
    sanitised_text: Optional[str] = None


# ─────────────────────────────────────────────────────────────────────────────
# Validation
# ─────────────────────────────────────────────────────────────────────────────

def validate_query(
    text: str,
    max_length: int = MAX_QUERY_LENGTH,
    check_injection: bool = True,
) -> ValidationResult:
    """Validate a user query before it enters the RAG pipeline.

    Returns ValidationResult with .valid=True and .sanitised_text set on success,
    or .valid=False with .reason explaining the rejection.
    """
    if not isinstance(text, str):
        return ValidationResult(valid=False, reason="Query must be a string")

    if not text.strip():
        return ValidationResult(valid=False, reason="Query is empty")

    if len(text) > max_length:
        return ValidationResult(
            valid=False,
            reason=f"Query exceeds {max_length} characters ({len(text)} given)",
        )

    if check_injection:
        for pattern in _INJECTION_PATTERNS:
            if pattern.search(text):
                return ValidationResult(
                    valid=False,
                    reason="Query matches a prompt-injection pattern",
                )

    # Basic sanitisation: collapse excessive whitespace, strip
    sanitised = re.sub(r"\s+", " ", text).strip()

    return ValidationResult(valid=True, sanitised_text=sanitised)

File: rust_output/src/input_guard.rs
/// Core/input_guard::py — Input validation for RAG queries.
/// 
/// Ported from main-app (George branch) E-12 input length guard.
/// Validates query length, detects basic prompt-injection patterns,
/// and ensures Unicode safety for Romanian/multilingual text.

use anyhow::{Result, Context};
use regex::Regex;
use serde::{Serialize, Deserialize};

pub const MAX_QUERY_LENGTH: i64 = 8000;

pub static _INJECTION_PATTERNS: std::sync::LazyLock<Vec<re::compile>> = std::sync::LazyLock::new(|| Vec::new());

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationResult {
    pub valid: bool,
    pub reason: Option<String>,
    pub sanitised_text: Option<String>,
}

/// Validate a user query before it enters the RAG pipeline.
/// 
/// Returns ValidationResult with .valid=true and .sanitised_text set on success,
/// or .valid=false with .reason explaining the rejection.
pub fn validate_query(text: String, max_length: i64, check_injection: bool) -> ValidationResult {
    // Validate a user query before it enters the RAG pipeline.
    // 
    // Returns ValidationResult with .valid=true and .sanitised_text set on success,
    // or .valid=false with .reason explaining the rejection.
    if !/* /* isinstance(text, str) */ */ true {
        ValidationResult(/* valid= */ false, /* reason= */ "Query must be a string".to_string())
    }
    if !text.trim().to_string() {
        ValidationResult(/* valid= */ false, /* reason= */ "Query is empty".to_string())
    }
    if text.len() > max_length {
        ValidationResult(/* valid= */ false, /* reason= */ format!("Query exceeds {} characters ({} given)", max_length, text.len()))
    }
    if check_injection {
        for pattern in _INJECTION_PATTERNS.iter() {
            if pattern.search(text) {
                ValidationResult(/* valid= */ false, /* reason= */ "Query matches a prompt-injection pattern".to_string())
            }
        }
    }
    let mut sanitised = regex::Regex::new(&"\\s+".to_string()).unwrap().replace_all(&" ".to_string(), text).to_string().trim().to_string();
    ValidationResult(/* valid= */ true, /* sanitised_text= */ sanitised)
}
